File: src/test/java/com/koddy/server/member/presentation/UpdateMentorProfileApiTest.java
package com.koddy.server.member.presentation;

import com.koddy.server.auth.exception.AuthExceptionCode;
import com.koddy.server.common.ControllerTest;
import com.koddy.server.member.application.usecase.UpdateMentorProfileUseCase;
import com.koddy.server.member.domain.model.Language;
import com.koddy.server.member.domain.model.mentee.Mentee;
import com.koddy.server.member.domain.model.mentor.Mentor;
import com.koddy.server.member.presentation.request.MentorScheduleRequest;
import com.koddy.server.member.presentation.request.UpdateMentorBasicInfoRequest;
import com.koddy.server.member.presentation.request.UpdateMentorScheduleRequest;
import com.koddy.server.member.presentation.request.model.LanguageRequestModel;
import com.koddy.server.member.presentation.request.model.MentoringPeriodRequestModel;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;

import java.time.LocalDate;
import java.util.List;

import static com.koddy.server.common.fixture.MenteeFixture.MENTEE_1;
import static com.koddy.server.common.fixture.MentorFixture.MENTOR_1;
import static com.koddy.server.common.utils.RestDocsSpecificationUtils.SnippetFactory.body;
import static com.koddy.server.common.utils.RestDocsSpecificationUtils.createHttpSpecSnippets;
import static com.koddy.server.common.utils.RestDocsSpecificationUtils.failureDocsWithAccessToken;
import static com.koddy.server.common.utils.RestDocsSpecificationUtils.successDocsWithAccessToken;
import static com.koddy.server.member.domain.model.mentor.DayOfWeek.MON;
import static com.koddy.server.member.domain.model.mentor.DayOfWeek.WED;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doNothing;
import static org.springframework.restdocs.payload.PayloadDocumentation.requestFields;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@DisplayName("Member -> UpdateMentorProfileApi 테스트")
class UpdateMentorProfileApiTest extends ControllerTest {
    @Autowired
    private UpdateMentorProfileUseCase updateMentorProfileUseCase;

    private final Mentor mentor = MENTOR_1.toDomain().apply(1L);
    private final Mentee mentee = MENTEE_1.toDomain().apply(2L);

    @Nested
    @DisplayName("멘토 기본정보 수정 API [PATCH /api/mentors/me/basic-info]")
    class UpdateBasicInfo {
        private static final String BASE_URL = "/api/mentors/me/basic-info";
        private final UpdateMentorBasicInfoRequest request = new UpdateMentorBasicInfoRequest(
                MENTOR_1.getName(),
                MENTOR_1.getProfileImageUrl(),
                MENTOR_1.getIntroduction(),
                new LanguageRequestModel(
                        Language.Category.KR.getCode(),
                        List.of(
                                Language.Category.EN.getCode(),
                                Language.Category.JP.getCode(),
                                Language.Category.CN.getCode()
                        )
                ),
                MENTOR_1.getUniversityProfile().getSchool(),
                MENTOR_1.getUniversityProfile().getMajor(),
                MENTOR_1.getUniversityProfile().getEnteredIn()
        );

        @Test
        @DisplayName("멘토가 아니면 권한이 없다")
        void throwExceptionByInvalidPermission() {
            // given
            applyToken(true, mentee);

            // when - then
            failedExecute(
                    patchRequestWithAccessToken(BASE_URL, request),
                    status().isForbidden(),
                    ExceptionSpec.of(AuthExceptionCode.INVALID_PERMISSION),
                    failureDocsWithAccessToken("MemberApi/Update/Mentor/BasicInfo/Failure", createHttpSpecSnippets(
                            requestFields(
                                    body("name", "이름", true),
                                    body("profileImageUrl", "프로필 이미지 URL", false),
                                    body("introduction", "멘토 자기소개", false),
                                    body("languages", "사용 가능한 언어", "KR EN CN JP VN", true),
                                    body("languages.main", "메인 언어", "1개", true),
                                    body("languages.sub[]", "서브 언어", "0..N개", false),
                                    body("school", "학교", true),
                                    body("major", "전공", true),
                                    body("enteredIn", "학번", true)
                            )
                    ))
            );
        }

        @Test
        @DisplayName("멘토 기본정보를 수정한다")
        void success() {
            // given
            applyToken(true, mentor);
            doNothing()
                    .when(updateMentorProfileUseCase)
                    .updateBasicInfo(any());

            // when - then
            successfulExecute(
                    patchRequestWithAccessToken(BASE_URL, request),
                    status().isNoContent(),
                    successDocsWithAccessToken("MemberApi/Update/Mentor/BasicInfo/Success", createHttpSpecSnippets(
                            requestFields(
                                    body("name", "이름", true),
                                    body("profileImageUrl", "프로필 이미지 URL", false),
                                    body("introduction", "멘토 자기소개", false),
                                    body("languages", "사용 가능한 언어", "KR EN CN JP VN", true),
                                    body("languages.main", "메인 언어", "1개", true),
                                    body("languages.sub[]", "서브 언어", "0..N개", false),
                                    body("school", "학교", true),
                                    body("major", "전공", true),
                                    body("enteredIn", "학번", true)
                            )
                    ))
            );
        }
    }

    @Nested
    @DisplayName("멘토 스케줄 수정 API [PATCH /api/mentors/me/schedules]")
    class UpdateSchedule {
        private static final String BASE_URL = "/api/mentors/me/schedules";
        private final UpdateMentorScheduleRequest request = new UpdateMentorScheduleRequest(
                new MentoringPeriodRequestModel(
                        LocalDate.of(2024, 1, 1),
                        LocalDate.of(2024, 5, 1)
                ),
                List.of(
                        new MentorScheduleRequest(MON.getKor(), new MentorScheduleRequest.Start(9, 0), new MentorScheduleRequest.End(17, 0)),
                        new MentorScheduleRequest(WED.getKor(), new MentorScheduleRequest.Start(13, 0), new MentorScheduleRequest.End(20, 0))
                )
        );

        @Test
        @DisplayName("멘토가 아니면 권한이 없다")
        void throwExceptionByInvalidPermission() {
            // given
            applyToken(true, mentee);

            // when - then
            failedExecute(
                    patchRequestWithAccessToken(BASE_URL, request),
                    status().isForbidden(),
                    ExceptionSpec.of(AuthExceptionCode.INVALID_PERMISSION),
                    failureDocsWithAccessToken("MemberApi/Update/Mentor/Schedule/Failure", createHttpSpecSnippets(
                            requestFields(
                                    body("period", "멘토링 시간 관련 설정", false),
                                    body("period.startDate", "멘토링 시작 날짜", "[KST] yyyy-MM-dd", false),
                                    body("period.endDate", "멘토링 종료 날짜", "[KST] yyyy-MM-dd", false),
                                    body("schedules", "멘토링 스케줄", false),
                                    body("schedules[].dayOfWeek", "날짜", "월 화 수 목 금 토 일", false),
                                    body("schedules[].start.hour", "시작 시간 (Hour)", "0 ~ 23", false),
                                    body("schedules[].start.minute", "시작 시간 (Minute)", "0 ~ 59", false),
                                    body("schedules[].end.hour", "종료 시간 (Hour)", "0 ~ 23", false),
                                    body("schedules[].end.minute", "종료 시간 (Minute)", "0 ~ 59", false)
                            )
                    ))
            );
        }

        @Test
        @DisplayName("멘토 스케줄을 수정한다")
        void success() {
            // given
            applyToken(true, mentor);
            doNothing()
                    .when(updateMentorProfileUseCase)
                    .updateSchedule(any());

            // when - then
            successfulExecute(
                    patchRequestWithAccessToken(BASE_URL, request),
                    status().isNoContent(),
                    successDocsWithAccessToken("MemberApi/Update/Mentor/Schedule/Success", createHttpSpecSnippets(
                            requestFields(
                                    body("period", "멘토링 시간 관련 설정", false),
                                    body("period.startDate", "멘토링 시작 날짜", "[KST] yyyy-MM-dd", false),
                                    body("period.endDate", "멘토링 종료 날짜", "[KST] yyyy-MM-dd", false),
                                    body("schedules", "멘토링 스케줄", false),
                                    body("schedules[].dayOfWeek", "날짜", "월 화 수 목 금 토 일", false),
                                    body("schedules[].start.hour", "시작 시간 (Hour)", "0 ~ 23", false),
                                    body("schedules[].start.minute", "시작 시간 (Minute)", "0 ~ 59", false),
                                    body("schedules[].end.hour", "종료 시간 (Hour)", "0 ~ 23", false),
                                    body("schedules[].end.minute", "종료 시간 (Minute)", "0 ~ 59", false)
                            )
                    ))
            );
        }
    }
}

File: src/test/java/com/koddy/server/member/presentation/UpdateMentorProfileApiTest.kt
package com.koddy.server.member.presentation

import com.koddy.server.auth.exception.AuthExceptionCode
import com.koddy.server.common.ApiDocsTestKt
import com.koddy.server.common.docs.ARRAY
import com.koddy.server.common.docs.DocumentField
import com.koddy.server.common.docs.NUMBER
import com.koddy.server.common.docs.OBJECT
import com.koddy.server.common.docs.STRING
import com.koddy.server.common.toLocalDate
import com.koddy.server.member.application.usecase.UpdateMentorProfileUseCase
import com.koddy.server.member.domain.model.Language
import com.koddy.server.member.domain.model.mentor.DayOfWeek
import com.koddy.server.member.presentation.request.MentorScheduleRequest
import com.koddy.server.member.presentation.request.UpdateMentorBasicInfoRequest
import com.koddy.server.member.presentation.request.UpdateMentorScheduleRequest
import com.koddy.server.member.presentation.request.model.LanguageRequestModel
import com.koddy.server.member.presentation.request.model.MentoringPeriodRequestModel
import com.ninjasquad.springmockk.MockkBean
import io.mockk.justRun
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest

@WebMvcTest(UpdateMentorProfileApi::class)
@DisplayName("Member -> UpdateMentorProfileApi 테스트")
internal class UpdateMentorProfileApiTest : ApiDocsTestKt() {
    @MockkBean
    private lateinit var updateMentorProfileUseCase: UpdateMentorProfileUseCase

    @Nested
    @DisplayName("멘토 기본정보 수정 API [PATCH /api/mentors/me/basic-info]")
    internal inner class UpdateBasicInfo {
        private val baseUrl = "/api/mentors/me/basic-info"
        private val request = UpdateMentorBasicInfoRequest(
            name = mentor.name,
            profileImageUrl = mentor.profileImageUrl,
            introduction = mentor.introduction,
            languages = LanguageRequestModel(
                main = Language.Category.KR.code,
                sub = listOf(
                    Language.Category.EN.code,
                    Language.Category.JP.code,
                    Language.Category.CN.code,
                ),
            ),
            school = mentor.universityProfile.school,
            major = mentor.universityProfile.major,
            enteredIn = mentor.universityProfile.enteredIn,
        )

        private val requestFields: Array<DocumentField> = arrayOf(
            "name" type STRING means "이름",
            "profileImageUrl" type STRING means "프로필 이미지 URL" isOptional true,
            "introduction" type STRING means "자기소개" isOptional true,
            "languages" type OBJECT means "사용 가능한 언어" constraint "KR EN CN JP VN",
            "languages.main" type STRING means "메인 언어" constraint "1개",
            "languages.sub[]" type ARRAY means "서브 언어" isOptional true,
            "school" type STRING means "학교",
            "major" type STRING means "전공",
            "enteredIn" type NUMBER means "학번",
        )

        @Test
        fun `멘토가 아니면 권한이 없다`() {
            val exceptionCode = AuthExceptionCode.INVALID_PERMISSION

            patchRequest(baseUrl) {
                accessToken(mentee)
                bodyContent(request)
            }.andExpect {
                status { isForbidden() }
                content { exception(exceptionCode) }
            }.andDo {
                makeFailureDocsWithAccessToken("MemberApi/Update/Mentor/BasicInfo/Failure") {
                    requestFields(*requestFields)
                }
            }
        }

        @Test
        fun `멘토 기본정보를 수정한다`() {
            justRun { updateMentorProfileUseCase.updateBasicInfo(any()) }

            patchRequest(baseUrl) {
                accessToken(mentor)
                bodyContent(request)
            }.andExpect {
                status { isNoContent() }
            }.andDo {
                makeSuccessDocsWithAccessToken("MemberApi/Update/Mentee/BasicInfo/Success") {
                    requestFields(*requestFields)
                }
            }
        }
    }

    @Nested
    @DisplayName("멘토 스케줄 수정 API [PATCH /api/mentors/me/schedules]")
    internal inner class UpdateSchedule {
        private val baseUrl = "/api/mentors/me/schedules"
        private val request = UpdateMentorScheduleRequest(
            period = MentoringPeriodRequestModel(
                startDate = "2024/01/01".toLocalDate(),
                endDate = "2024/05/01".toLocalDate(),
            ),
            schedules = listOf(
                MentorScheduleRequest(
                    dayOfWeek = DayOfWeek.MON.kor,
                    start = MentorScheduleRequest.Start(9, 0),
                    end = MentorScheduleRequest.End(17, 0),
                ),
                MentorScheduleRequest(
                    dayOfWeek = DayOfWeek.WED.kor,
                    start = MentorScheduleRequest.Start(13, 0),
                    end = MentorScheduleRequest.End(20, 0),
                ),
            ),
        )

        private val requestFields: Array<DocumentField> = arrayOf(
            "period" type OBJECT means "멘토링 기간" isOptional true,
            "period.startDate" type STRING means "멘토링 기간 [시작 날짜]" constraint "[KST] yyyy-MM-dd",
            "period.endDate" type STRING means "멘토링 기간 [종료 날짜]" constraint "[KST] yyyy-MM-dd",
            "schedules" type ARRAY means "멘토링 주간 스케줄" isOptional true,
            "schedules[].dayOfWeek" type STRING means "멘토링 주간 스케줄 [요일]",
            "schedules[].start.hour" type NUMBER means "멘토링 주간 스케줄 [시작 시간(Hour)]" constraint "0 ~ 23",
            "schedules[].start.minute" type NUMBER means "멘토링 주간 스케줄 [종료 시간(Minute)]" constraint "0 ~ 59",
            "schedules[].end.hour" type NUMBER means "멘토링 주간 스케줄 [시작 시간(Hour)]" constraint "0 ~ 23",
            "schedules[].end.minute" type NUMBER means "멘토링 주간 스케줄 [종료 시간(Minute)]" constraint "0 ~ 59",
        )

        @Test
        fun `멘토가 아니면 권한이 없다`() {
            val exceptionCode = AuthExceptionCode.INVALID_PERMISSION

            patchRequest(baseUrl) {
                accessToken(mentee)
                bodyContent(request)
            }.andExpect {
                status { isForbidden() }
                content { exception(exceptionCode) }
            }.andDo {
                makeFailureDocsWithAccessToken("MemberApi/Update/Mentor/Schedule/Failure") {
                    requestFields(*requestFields)
                }
            }
        }

        @Test
        fun `멘토 스케줄을 수정한다`() {
            justRun { updateMentorProfileUseCase.updateSchedule(any()) }

            patchRequest(baseUrl) {
                accessToken(mentor)
                bodyContent(request)
            }.andExpect {
                status { isNoContent() }
            }.andDo {
                makeSuccessDocsWithAccessToken("MemberApi/Update/Mentor/Schedule/Success") {
                    requestFields(*requestFields)
                }
            }
        }
    }
}
